File: src/nl/rubensten/texifyidea/run/LatexRunConfigurationProducer.java
package nl.rubensten.texifyidea.run;

import com.intellij.execution.Location;
import com.intellij.execution.actions.ConfigurationContext;
import com.intellij.execution.actions.RunConfigurationProducer;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.util.Ref;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import nl.rubensten.texifyidea.file.LatexFileType;
import nl.rubensten.texifyidea.util.PsiUtilKt;

/**
 * @author Ruben Schellekens
 */
public class LatexRunConfigurationProducer extends RunConfigurationProducer<LatexRunConfiguration> {

    protected LatexRunConfigurationProducer() {
        super(LatexRunConfigurationType.getInstance());
    }

    @Override
    protected boolean setupConfigurationFromContext(LatexRunConfiguration runConfiguration,
                                                    ConfigurationContext context,
                                                    Ref<PsiElement> sourceElement) {
        Location location = context.getLocation();
        if (location == null) {
            return false;
        }

        PsiFile container = getEntryPointContainer(location);
        if (container == null) {
            return false;
        }

        VirtualFile mainFile = container.getVirtualFile();
        if (mainFile == null) {
            return false;
        }

        // Only activate on .tex files.
        final String extension = mainFile.getExtension();
        final String texTension = LatexFileType.INSTANCE.getDefaultExtension();
        if (extension == null || !extension.equalsIgnoreCase(texTension)) {
            return false;
        }

        // Setup run configuration.
        runConfiguration.setMainFile(mainFile);
        runConfiguration.setDefaultAuxiliaryDirectories();
        runConfiguration.setDefaultCompiler();
        runConfiguration.setDefaultOutputFormat();
        runConfiguration.setSuggestedName();

        if (PsiUtilKt.hasBibliography(container)) {
            runConfiguration.generateBibRunConfig();
        }

        return true;
    }

    private PsiFile getEntryPointContainer(Location location) {
        if (location == null) {
            return null;
        }

        PsiElement locationElement = location.getPsiElement();
        return locationElement.getContainingFile();
    }

    @Override
    public boolean isConfigurationFromContext(LatexRunConfiguration runConfiguration,
                                              ConfigurationContext context) {
        VirtualFile mainFile = runConfiguration.getMainFile();
        PsiFile psiFile = context.getDataContext().getData(PlatformDataKeys.PSI_FILE);
        if (psiFile == null) {
            return false;
        }

        VirtualFile currentFile = psiFile.getVirtualFile();

        return mainFile.getPath().equals(currentFile.getPath());
    }
}

File: src/nl/rubensten/texifyidea/run/LatexRunConfigurationProducer.kt
package nl.rubensten.texifyidea.run

import com.intellij.execution.Location
import com.intellij.execution.actions.ConfigurationContext
import com.intellij.execution.actions.RunConfigurationProducer
import com.intellij.openapi.actionSystem.PlatformDataKeys
import com.intellij.openapi.util.Ref
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import nl.rubensten.texifyidea.file.LatexFileType
import nl.rubensten.texifyidea.util.hasBibliography

/**
 * @author Ruben Schellekens
 */
class LatexRunConfigurationProducer : RunConfigurationProducer<LatexRunConfiguration>(LatexRunConfigurationType.instance) {

    override fun setupConfigurationFromContext(runConfiguration: LatexRunConfiguration,
                                               context: ConfigurationContext,
                                               sourceElement: Ref<PsiElement>): Boolean {
        val location = context.location ?: return false
        val container = getEntryPointContainer(location) ?: return false
        val mainFile = container.virtualFile ?: return false

        // Only activate on .tex files.
        val extension = mainFile.extension
        val texTension = LatexFileType.defaultExtension
        if (extension == null || !extension.equals(texTension, ignoreCase = true)) {
            return false
        }

        // Setup run configuration.
        runConfiguration.mainFile = mainFile
        runConfiguration.setDefaultAuxiliaryDirectories()
        runConfiguration.setDefaultCompiler()
        runConfiguration.setDefaultOutputFormat()
        runConfiguration.setSuggestedName()

        if (container.hasBibliography()) {
            runConfiguration.generateBibRunConfig()
        }

        return true
    }

    private fun getEntryPointContainer(location: Location<*>?): PsiFile? {
        if (location == null) {
            return null
        }

        val locationElement = location.psiElement
        return locationElement.containingFile
    }

    override fun isConfigurationFromContext(runConfiguration: LatexRunConfiguration,
                                            context: ConfigurationContext): Boolean {
        val mainFile = runConfiguration.mainFile
        val psiFile = context.dataContext.getData(PlatformDataKeys.PSI_FILE) ?: return false

        val currentFile = psiFile.virtualFile

        return mainFile.path == currentFile.path
    }
}